src/ToData.purs

```diff
@@ -1,38 +1,138 @@
 module ToData
   ( class ToData
+  , class ToDataArgs
+  , class ToDataWithIndex
+  , class ToDataArgsRL
+  , genericToData
+  , toDataArgsRec
   , toData
+  , toDataArgs
+  , toDataWithIndex
   ) where
 
 import Prelude
 
+import ConstrIndex (class HasConstrIndex, constrIndex)
+import Data.Array (cons)
 import Data.Array as Array
 import Data.BigInt (BigInt)
 import Data.BigInt as BigInt
 import Data.Either (Either(Left, Right))
 import Data.Foldable (class Foldable)
+import Data.Generic.Rep as G
 import Data.List (List)
 import Data.Map (Map)
 import Data.Map as Map
 import Data.Maybe (Maybe(Just, Nothing))
 import Data.Profunctor.Strong ((***))
 import Data.Ratio (Ratio, denominator, numerator)
+import Data.Symbol (class IsSymbol, SProxy(..), reflectSymbol)
 import Data.Tuple (Tuple(Tuple))
-import Data.Tuple.Nested (type (/\))
+import Data.Tuple.Nested (type (/\), (/\))
 import Data.UInt (UInt)
 import Helpers (uIntToBigInt)
+import Prim.Row as Row
+import Prim.RowList as RL
 import Prim.TypeError (class Fail, Text)
+import Record as Record
+import Type.Proxy (Proxy(Proxy))
 import Types.ByteArray (ByteArray)
 import Types.PlutusData (PlutusData(Constr, Integer, List, Map, Bytes))
 
+-- | Classes
+
+class ToData :: Type -> Constraint
 class ToData (a :: Type) where
   toData :: a -> PlutusData
 
+class ToDataWithIndex :: Type -> Type -> Constraint
+class ToDataWithIndex a ci where
+  toDataWithIndex :: HasConstrIndex ci => Proxy ci -> a -> PlutusData
+
+-- As explained in https://harry.garrood.me/blog/write-your-own-generics/ this
+-- is just a neat pattern that flattens a skewed Product of Products
+class ToDataArgs :: Type -> Constraint
+class ToDataArgs a where
+  toDataArgs :: a -> Array (PlutusData)
+
+-- | A helper typeclass to implement `ToDataArgs` for records.
+-- Stolen from https://github.com/purescript/purescript-quickcheck/blob/v7.1.0/src/Test/QuickCheck/Arbitrary.purs#L247
+class ToDataArgsRL :: RL.RowList Type -> Row Type -> Constraint
+class ToDataArgsRL list row | list -> row where
+  toDataArgsRec :: forall rlproxy. rlproxy list -> Record row -> Array PlutusData
+
+-- | ToDataWithIndex instances for Data.Generic.Rep
+
+instance (HasConstrIndex a, ToDataWithIndex l a, ToDataWithIndex r a) => ToDataWithIndex (G.Sum l r) a where
+  toDataWithIndex p (G.Inl x) = toDataWithIndex p x
+  toDataWithIndex p (G.Inr x) = toDataWithIndex p x
+
+instance (IsSymbol n, HasConstrIndex a, ToDataArgs arg) => ToDataWithIndex (G.Constructor n arg) a where
```

Comment:
My previous review was potentially a bit confusing. But basically, if we decide to keep the constraint implicit, we could write this:
```
instance (ToDataWithIndex l a, ToDataWithIndex r a) => ToDataWithIndex (G.Sum l r) a where
  toDataWithIndex p (G.Inl x) = toDataWithIndex p x
  toDataWithIndex p (G.Inr x) = toDataWithIndex p x

instance (IsSymbol n, ToDataArgs arg) => ToDataWithIndex (G.Constructor n arg) a where
  toDataWithIndex p (G.Constructor args) = Constr (resolveIndex p (SProxy :: SProxy n)) (toDataArgs args)
```
so symmetry is preserved anyhow.

Reply:
I think it worked out very nicely!